### ISSUE TITLE
Visualise microservice integrations / connections

### ISSUE THREAD

**Issue by rawlingsj** [fabric8io/fabric8#4228]
A number of people have raised an issue with working with microservices is it's difficult to know the relationships and dependencies between their services.

It would be great if we could visualise a microservices architecture using fabric8 related tech such as the console, API Man and Camel when used.

**Comment by EricWittmann** [comment]
At design time apiman can tell which services can be called by which applications.  At runtime we would need to invent something new.  We're in a good position to do it though - assuming service contracts are always used.  If services are marked as public, then there is no way to know which client is making a call.

Note: a similar problem that the rtgov project attempts to solve is tracking a business transaction across multiple rest calls.  I'm not sure how it does the correlation, presumably that needs to be configured per service, since the correlation data may come from various places (header, resource path, request body).

**Comment by KurtStam** [comment]
I talked to Ioannis last week about this and he created some sort of dependency mechanism where he looks into dependent jars if they have a kubernetes.json file in them. If they do then it is captured in kubernetes file and when the app is started using fabric8:apply it checks if that app is running. Some similar check could happen when starting from the console. In the case of apiman depending on elastic search we'd have to add elastic as a 'provided' dependency. If we can capture the service dependencies then we could visualize them in the console too.

**Comment by jimmidyson** [comment]
I think the only way to ensure that correlation IDs are propagated by clients is to instrument via a java agent & cover as much stuff as we can. I can't think of another way to do it tbh without correlation IDs in headers, etc being dropped by clients.

**Comment by KurtStam** [comment]
There are no hooks in the kubernetes service to do this? That really seems to be the prefect place for things like these.

**Comment by jimmidyson** [comment]
No there are no hooks in the kubernetes service layer to do this. Even then it's still up to the client to send the correlation ID or not, I'd assume service layer would just add a new one if it didn't exist (if this were actually possible).

**Comment by EricWittmann** [comment]
The other option is custom correlation policies in apiman, configured for each service.  This allows extracting correlation information from wherever it makes sense (custom headers, resource path, payload, etc).  But it requires service providers to define the information manually.  You end up using something service/business specific for the correlation (e.g. invoiceId).

Also no reason we couldn't allow both approaches.

**Comment by KurtStam** [comment]
The client side problem is rather easy I think - we can simply enforce it being send it :).

**Comment by jimmidyson** [comment]
How?

**Comment by KurtStam** [comment]
I meant 'require'. Just document that that you are supposed to send in a required header. Then in the gateway or elsewhere we can fail the request if it doesn't have it.

**Comment by jimmidyson** [comment]
Yeah we could but it only takes one to drop it & we lose all correlation. Also that requires code changes which could be so much easier if we can just instrument which also gives us the opportunity to add/remove stuff as we need without apps having to do that themselves.
